cpp/src/arrow/type.h

```diff
@@ -250,15 +251,55 @@ struct ListType : public DataType {
   static char const* name() { return "list"; }
 
   std::string ToString() const override;
+
+ protected:
+  // Constructor for classes that are implemented as List Arrays.
+  ListType(const std::shared_ptr<DataType>& value_type, Type::type logical_type)
+      : DataType(logical_type) {
+    // TODO ARROW-187 this can technically fail, make a constructor method ?
+    children_ = {std::make_shared<Field>("item", value_type)};
+  }
+};
+
+// BinaryType type is reprsents lists of 1-byte values.
+struct BinaryType : public ListType {
```

Comment:
One thing we might examine in some detail is: to what extent do we benefit from Binary/String being semantically a subclass of ListType (and therefore, technically a nested type)? This is orthogonal to the memory layout issue and strictly a metadata matter. I can see an argument for having a class hierarchy of String > Binary > Primitive, and we would permit copyless casting between List<UInt8> and these types